### PR TITLE
SONARPY-869 Deserialize protobuf files containing symbol information …

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
…for "annoy" library

### CODE REVIEW

File: python-frontend/pom.xml [hunk 1]
```diff
@@ -37,6 +37,68 @@
...
       <artifactId>guava</artifactId>
     </dependency>
+    <dependency>
+      <groupId>com.google.protobuf</groupId>
+      <artifactId>protobuf-java</artifactId>
+      <version>${protobuf.version}</version>
+    </dependency>
   </dependencies>
-
+  <build>
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/pom.xml:46]: Just to make sure, this is the same thing we do in sonar-security to compile the Java Protobuf classes automatically, right?
Will it need adaptations to build the Python classes as well? (Should we even do it here as well?)

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 2]
```diff
@@ -340,4 +354,77 @@ private static void setParameterTypes(Symbol symbol, FunctionDef functionDef) {
...
+    if (fqnSplittedByDot.length == 2 && fqnSplittedByDot[0].equals("builtins") && builtins.containsKey(localName)) {
+      return builtins.get(localName);
+    }
+    for (Map<String, Symbol> symbolsByFqn : typeShedSymbols.values()) {
+      Symbol symbol = symbolsByFqn.get(fullyQualifiedName);
+      if (symbol != null) {
+        return symbol;
+      }
+    }
+    return builtins.getOrDefault(fullyQualifiedName, null);
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:427]: Can it happen that we actually retrieve a builtin here if the name didn't start with `builtins`? (for serialized Typeshed at least)

File: python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java [hunk 3]
```diff
@@ -154,6 +155,30 @@ public static InferredType fromTypeshedTypeAnnotation(TypeAnnotation typeAnnotat
...
   }
 
+  public static InferredType fromTypeshedProtobuf(@Nullable SymbolsProtos.Type type) {
+    if (type == null) {
+      return anyType();
+    }
+    switch (type.getKind()) {
+      case INSTANCE:
+        String typeName = type.getSimpleName();
+        return typeName != null ? runtimeType(TypeShed.symbolWithFQN(typeName)) : anyType();
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java:165]: Here as well it seems `typeName` won't be null.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Overall looks good. Left a few comments and clarifying questions. I played a bit the devil's advocate and I'm not sure everything needs or can actually be addressed (especially the part that relates to the handling of super classes).

File: python-frontend/src/test/java/org/sonar/python/types/TypeShedTest.java [hunk 4]
```diff
@@ -234,4 +244,140 @@ public void stub_files_symbols() {
...
+  @Test
+  public void deserialize_annoy_protobuf() {
+    Map<String, Symbol> deserializedAnnoySymbols = TypeShed.symbolsForModule("annoy").stream()
+      .collect(Collectors.toMap(Symbol::fullyQualifiedName, s -> s));
+    assertThat(deserializedAnnoySymbols.values()).extracting(Symbol::kind, Symbol::fullyQualifiedName)
+      .containsExactlyInAnyOrder(tuple(Kind.CLASS, "annoy._Vector"), tuple(Kind.CLASS, "annoy.AnnoyIndex"));
+
+    ClassSymbol vector = (ClassSymbol) deserializedAnnoySymbols.get("annoy._Vector");
+    assertThat(vector.superClasses()).extracting(Symbol::kind, Symbol::fullyQualifiedName)
+      .containsExactlyInAnyOrder(tuple(Kind.AMBIGUOUS, "typing.Sized"));
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/test/java/org/sonar/python/types/TypeShedTest.java:257]: Yes that's correct

File: sonar-python-plugin/pom.xml [hunk 5]
```diff
@@ -150,8 +150,8 @@
             <configuration>
               <rules>
                 <requireFilesSize>
-                  <maxsize>5000000</maxsize>
-                  <minsize>4500000</minsize>
+                  <maxsize>6500000</maxsize>
+                  <minsize>6000000</minsize>
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply sonar-python-plugin/pom.xml:154]: Yes, this is because of protobuf depedency which takes around 1.6MB

File: python-frontend/pom.xml (hunk 1 above)
Comment (andrea-guarino-sonarsource) [reply python-frontend/pom.xml:46]: Yes, that's the same as sonar-security. I believe it makes sense to also build the Python classes here, either on this pom or in a dedicated pom if we decide to have a separate module

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java (hunk 2 above)
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:427]: I think you're right, it should never happen for serialized Typeshed

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 6]
```diff
@@ -340,4 +354,77 @@ private static void setParameterTypes(Symbol symbol, FunctionDef functionDef) {
...
+    }
+    for (SymbolsProtos.OverloadedFunctionSymbol overloadedFunctionSymbol : moduleSymbol.getOverloadedFunctionsList()) {
+      deserializedSymbols.put(overloadedFunctionSymbol.getFullname(),
+        AmbiguousSymbolImpl.create(overloadedFunctionSymbol.getDefinitionsList().stream().map(FunctionSymbolImpl::new).collect(Collectors.toSet())));
+    }
+    return deserializedSymbols;
+  }
+
+
+  private static void addSuperClasses(Map<String, Symbol> deserializedSymbols, SymbolsProtos.ClassSymbol classSymbolProto, ClassSymbolImpl classSymbol) {
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:397]: I think it makes sense! Thanks for the nice suggestion!

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**